File: csharp/Database/Revenj.DatabasePersistence.Postgres/Converters/RecordTuple.cs
using System;
using System.IO;
using Revenj.Utility;

namespace Revenj.DatabasePersistence.Postgres.Converters
{
	public class RecordTuple : IPostgresTuple
	{
		public static readonly IPostgresTuple Empty;
		public static readonly IPostgresTuple Null;

		static RecordTuple()
		{
			Empty = new EmptyRecordTuple();
			Null = new NullTuple();
		}

		private readonly IPostgresTuple[] Properties;

		internal RecordTuple(IPostgresTuple[] properties)
		{
			this.Properties = properties;
		}

		public static IPostgresTuple From(IPostgresTuple[] properties)
		{
			if (properties == null)
				return Null;
			if (properties.Length == 0)
				return Empty;
			return new RecordTuple(properties);
		}

		class EmptyRecordTuple : IPostgresTuple
		{
			public bool MustEscapeRecord { get { return true; } }
			public bool MustEscapeArray { get { return false; } }
			public void InsertRecord(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
			{
				sw.Write("()");
			}
			public void InsertArray(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
			{
				sw.Write("()");
			}
			public string BuildTuple(bool quote) { return quote ? "'()'" : "()"; }
		}

		class NullTuple : IPostgresTuple
		{
			public bool MustEscapeRecord { get { return false; } }
			public bool MustEscapeArray { get { return false; } }
			public void InsertRecord(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings) { }
			public void InsertArray(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
			{
				sw.Write("NULL");
			}
			public string BuildTuple(bool quote) { return "NULL"; }
		}

		public bool MustEscapeRecord { get { return true; } }
		public bool MustEscapeArray { get { return true; } }

		public string BuildTuple(bool quote)
		{
			using (var cms = ChunkedMemoryStream.Create())
			{
				var sw = cms.GetWriter();
				Action<TextWriter, char> mappings = null;
				if (quote)
				{
					mappings = PostgresTuple.EscapeQuote;
					sw.Write('\'');
				}
				sw.Write('(');
				var p = Properties[0];
				if (p != null)
				{
					if (p.MustEscapeRecord)
					{
						sw.Write('"');
						p.InsertRecord(sw, cms.SmallBuffer, "1", mappings);
						sw.Write('"');
					}
					else p.InsertRecord(sw, cms.SmallBuffer, string.Empty, mappings);
				}
				for (int i = 1; i < Properties.Length; i++)
				{
					sw.Write(',');
					p = Properties[i];
					if (p != null)
					{
						if (p.MustEscapeRecord)
						{
							sw.Write('"');
							p.InsertRecord(sw, cms.SmallBuffer, "1", mappings);
							sw.Write('"');
						}
						else p.InsertRecord(sw, cms.SmallBuffer, string.Empty, mappings);
					}
				}
				sw.Write(')');
				if (quote)
					sw.Write('\'');
				sw.Flush();
				cms.Position = 0;
				return cms.GetReader().ReadToEnd();
			}
		}

		private static readonly byte[] NULL = new byte[] { (byte)'N', (byte)'U', (byte)'L', (byte)'L' };

		public void InsertRecord(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
		{
			sw.Write('(');
			var newEscaping = escaping + '1';
			string quote = null;
			var p = Properties[0];
			if (p != null)
			{
				if (p.MustEscapeRecord)
				{
					quote = PostgresTuple.BuildQuoteEscape(escaping);
					if (mappings != null)
						foreach (var q in quote)
							mappings(sw, q);
					else
						sw.Write(quote);
					p.InsertRecord(sw, buf, newEscaping, mappings);
					if (mappings != null)
						foreach (var q in quote)
							mappings(sw, q);
					else
						sw.Write(quote);
				}
				else p.InsertRecord(sw, buf, escaping, mappings);
			}
			for (int i = 1; i < Properties.Length; i++)
			{
				sw.Write(',');
				p = Properties[i];
				if (p != null)
				{
					if (p.MustEscapeRecord)
					{
						//TODO: build quote only once and reuse it, instead of looping all the time
						quote = quote ?? PostgresTuple.BuildQuoteEscape(escaping);
						if (mappings != null)
							foreach (var q in quote)
								mappings(sw, q);
						else
							sw.Write(quote);
						p.InsertRecord(sw, buf, newEscaping, mappings);
						if (mappings != null)
							foreach (var q in quote)
								mappings(sw, q);
						else
							sw.Write(quote);
					}
					else p.InsertRecord(sw, buf, escaping, mappings);
				}
			}
			sw.Write(')');
		}

		public void InsertArray(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
		{
			InsertRecord(sw, buf, escaping, mappings);
		}
	}
}

File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/RecordTuple.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresWriter;

public class RecordTuple extends PostgresTuple {
	public static final PostgresTuple EMPTY;
	public static final PostgresTuple NULL;

	static {
		EMPTY = new EmptyRecordTuple();
		NULL = new NullTuple();
	}

	private final PostgresTuple[] properties;

	RecordTuple(PostgresTuple[] properties) {
		this.properties = properties;
	}

	public static PostgresTuple from(PostgresTuple[] properties) {
		if (properties == null) return NULL;
		if (properties.length == 0) return EMPTY;
		return new RecordTuple(properties);
	}

	private static class EmptyRecordTuple extends PostgresTuple {
		public boolean mustEscapeRecord() {
			return true;
		}

		public boolean mustEscapeArray() {
			return false;
		}

		public void insertRecord(PostgresWriter sw, String escaping, Mapping mappings) {
			sw.write("()");
		}

		public void insertArray(PostgresWriter sw, String escaping, Mapping mappings) {
			sw.write("()");
		}

		public String buildTuple(boolean quote) {
			return quote ? "'()'" : "()";
		}
	}

	private static class NullTuple extends PostgresTuple {
		public boolean mustEscapeRecord() {
			return false;
		}

		public boolean mustEscapeArray() {
			return false;
		}

		public void insertRecord(PostgresWriter sw, String escaping, Mapping mappings) {
		}

		public void insertArray(PostgresWriter sw, String escaping, Mapping mappings) {
			sw.write("NULL");
		}

		public String buildTuple(boolean quote) {
			return "NULL";
		}
	}

	public boolean mustEscapeRecord() {
		return true;
	}

	public boolean mustEscapeArray() {
		return true;
	}

	public String buildTuple(boolean quote) {
		PostgresWriter sw = threadWriter.get();
		sw.reset();
		Mapping mappings = null;
		if (quote) {
			mappings = PostgresTuple::escapeQuote;
			sw.write('\'');
		}
		sw.write('(');
		PostgresTuple p = properties[0];
		if (p != null) {
			if (p.mustEscapeRecord()) {
				sw.write('"');
				p.insertRecord(sw, "1", mappings);
				sw.write('"');
			} else p.insertRecord(sw, "", mappings);
		}
		for (int i = 1; i < properties.length; i++) {
			sw.write(',');
			p = properties[i];
			if (p != null) {
				if (p.mustEscapeRecord()) {
					sw.write('"');
					p.insertRecord(sw, "1", mappings);
					sw.write('"');
				} else p.insertRecord(sw, "", mappings);
			}
		}
		sw.write(')');
		if (quote) {
			sw.write('\'');
		}
		return sw.toString();
	}

	public void insertRecord(PostgresWriter sw, String escaping, Mapping mappings) {
		sw.write('(');
		String newEscaping = escaping + '1';
		String quote = null;
		PostgresTuple p = properties[0];
		if (p != null) {
			if (p.mustEscapeRecord()) {
				quote = buildQuoteEscape(escaping);
				if (mappings != null) {
					for (int x = 0; x < quote.length(); x++) {
						mappings.map(sw, quote.charAt(x));
					}
				} else sw.write(quote);
				p.insertRecord(sw, newEscaping, mappings);
				if (mappings != null) {
					for (int x = 0; x < quote.length(); x++) {
						mappings.map(sw, quote.charAt(x));
					}
				} else sw.write(quote);
			} else p.insertRecord(sw, escaping, mappings);
		}
		for (int i = 1; i < properties.length; i++) {
			sw.write(',');
			p = properties[i];
			if (p != null) {
				if (p.mustEscapeRecord()) {
					//TODO: build quote only once and reuse it, instead of looping all the time
					quote = quote != null ? quote : buildQuoteEscape(escaping);
					if (mappings != null) {
						for (int x = 0; x < quote.length(); x++) {
							mappings.map(sw, quote.charAt(x));
						}
					} else
						sw.write(quote);
					p.insertRecord(sw, newEscaping, mappings);
					if (mappings != null) {
						for (int x = 0; x < quote.length(); x++) {
							mappings.map(sw, quote.charAt(x));
						}
					} else sw.write(quote);
				} else p.insertRecord(sw, escaping, mappings);
			}
		}
		sw.write(')');
	}
}
